libparsec/crates/client/src/certif/user_update_profile.rs

```diff
@@ -0,0 +1,162 @@
+    #[error(transparent)]
+    InvalidCertificate(#[from] Box<InvalidCertificateError>),
+    #[error(transparent)]
+    Internal(#[from] anyhow::Error),
+}
+
+impl From<ConnectionError> for CertifUpdateUserProfileError {
+    fn from(value: ConnectionError) -> Self {
+        match value {
+            ConnectionError::NoResponse(_) => Self::Offline,
+            // TODO: handle organization expired and user revoked here ?
+            err => Self::Internal(err.into()),
+        }
+    }
+}
+
+impl From<CertifStoreError> for CertifUpdateUserProfileError {
+    fn from(value: CertifStoreError) -> Self {
+        match value {
+            CertifStoreError::Stopped => Self::Stopped,
+            CertifStoreError::Internal(err) => err.into(),
+        }
+    }
+}
+
+pub(super) async fn update_profile(
+    ops: &CertificateOps,
+    user: UserID,
+    new_profile: UserProfile,
+) -> Result<CertificateBasedActionOutcome, CertifUpdateUserProfileError> {
+    // Loop is needed to deal with server requiring greater timestamp
+    let mut timestamp = ops.device.now();
+    loop {
+        let outcome = do_server_command(ops, user, timestamp, new_profile).await?;
+
+        match outcome {
+            DoServerCommandOutcome::Done(outcome) => return Ok(outcome),
+            DoServerCommandOutcome::RequireGreaterTimestamp(strictly_greater_than) => {
+                // TODO: handle `strictly_greater_than` out of the client ballpark by
+                // returning an error
+                timestamp = greater_timestamp(
+                    &ops.device.time_provider,
+                    GreaterTimestampOffset::User,
+                    strictly_greater_than,
+                );
+            }
+        }
+    }
+}
+
+enum DoServerCommandOutcome {
+    Done(CertificateBasedActionOutcome),
+    RequireGreaterTimestamp(DateTime),
+}
+
+async fn do_server_command(
+    ops: &CertificateOps,
+    recipient: UserID,
+    timestamp: DateTime,
+    new_profile: UserProfile,
+) -> Result<DoServerCommandOutcome, CertifUpdateUserProfileError> {
+    // 1) Build update user profile certificate
+
+    let signed_certificate = UserUpdateCertificate {
+        author: ops.device.device_id,
+        timestamp,
+        user_id: recipient,
+        new_profile,
+    }
+    .dump_and_sign(&ops.device.signing_key);
+
+    // 2) Actually send the command to the server
+
+    use authenticated_cmds::latest::user_update::{Rep, Req};
+
+    let req = Req {
+        user_update_certificate: signed_certificate.into(),
+    };
+    let rep = ops.cmds.send(req).await?;
+    match rep {
+        Rep::Ok => Ok(DoServerCommandOutcome::Done(
+            CertificateBasedActionOutcome::Uploaded {
+                certificate_timestamp: timestamp,
+            },
+        )),
+        Rep::UserNoChanges => Ok(DoServerCommandOutcome::Done(
+            CertificateBasedActionOutcome::RemoteIdempotent {
+                certificate_timestamp: timestamp,
+            },
+        )),
```

Comment:
I couldn't find a test for this case, I missed something?

Reply:
I added it in the ok test